File: java/src/test/java/cz/zcu/pperncka/algorithm/FindFlowDeltaTest.java
package cz.zcu.pperncka.algorithm;

import cz.zcu.pperncka.data.Bush;
import cz.zcu.pperncka.data.Network;
import cz.zcu.pperncka.data.Network.Arc;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

@SuppressWarnings("unchecked")
@DisplayName("findFlowDelta")
public class FindFlowDeltaTest {
	
	static Algorithm a;
	
	@BeforeAll
	static void beforeAll() {
		Network n = new Network(new List[0], 0);
		a = new Algorithm(n, null, 0, 0);
	}
	
	@DisplayName("Case from Wikipedia - diamond")
	@Test
	void test1() {
		var arc01 = new Arc(0, 1, 	 1,  0, 8000, 0);
		var arc02 = new Arc(0, 2, 	 1,  0,    0, 0);
		var arc13 = new Arc(1, 3, 1000, 15, 8000, 0);
		var arc23 = new Arc(2, 3, 3000, 20,    0, 0);
		
		var minTree = new Arc[]{null, arc01, arc02, arc23};
		var maxTree = new Arc[]{null, arc01, arc02, arc13};
		
		Bush bush = new Bush(1, 0);
		bush.addFlow(0, 10000);
		
		assertEquals(5847, Math.round(a.findFlowDelta(minTree, maxTree, bush, 3, 0)));
	}
	
	@DisplayName("Case from Wikipedia - diamond with tail")
	@Test
	void test2() {
		var arc01 = new Arc(0, 1, 	 1,  0,    0, 0);
		var arc12 = new Arc(1, 2, 	 1,  0, 8000, 0);
		var arc13 = new Arc(1, 3, 	 1,  0,    0, 0);
		var arc24 = new Arc(2, 4, 1000, 15, 8000, 0);
		var arc34 = new Arc(3, 4, 3000, 20,    0, 0);
		
		var minTree = new Arc[]{null, arc01, arc12, arc13, arc34};
		var maxTree = new Arc[]{null, arc01, arc12, arc13, arc24};
		
		Bush bush = new Bush(1, 0);
		bush.addFlow(0, 10000);
		
		assertEquals(5847, Math.round(a.findFlowDelta(minTree, maxTree, bush, 4, 1)));
	}
	
	@DisplayName("Test maxDeltaX <= μ")
	@Test
	void test3() {
		var arc01 = new Arc(0, 1, 	 1,  0, 8000, 0);
		var arc02 = new Arc(0, 2, 	 1,  0,    0, 0);
		var arc13 = new Arc(1, 3, 1000, 15, 8000, 0);
		var arc23 = new Arc(2, 3, 3000, 20,    0, 0);
		
		var minTree = new Arc[]{null, arc01, arc02, arc23};
		var maxTree = new Arc[]{null, arc01, arc02, arc13};
		
		Bush bush = new Bush(1, 0);
		bush.addFlow(0, 1000);
		
		assertEquals(1000, a.findFlowDelta(minTree, maxTree, bush, 3, 0));
	}
	
	@DisplayName("Nonconverging case")
	@Test
	void test4() {
		var arc01 = new Arc(0, 1,  17782.7941, 2, 1900, 0);
		var arc12 = new Arc(1, 2,  4908.82673, 6,  700, 0);
		var arc23 = new Arc(2, 3, 		10000, 5, 20783.42548485832, 0);
		var arc04 = new Arc(0, 4, 		10000, 5, 2300, 0);
		var arc43 = new Arc(4, 3, 13915.78842, 3, 1900, 0);
		
		var minTree = new Arc[]{null, arc01, arc12, arc23, arc04};
		var maxTree = new Arc[]{null, arc01, arc12, arc43, arc04};
		
		Bush bush = new Bush(1, 0);
		bush.addFlow(0, 700);
		
		assertEquals(700, a.findFlowDelta(minTree, maxTree, bush, 3, 0));
	}
	
	@DisplayName("maxDeltaX == 0")
	@Test
	void test5() {
		var arc01 = new Arc(0, 1, 	 1,  0, 8000, 0);
		var arc02 = new Arc(0, 2, 	 1,  0,    0, 0);
		var arc13 = new Arc(1, 3, 1000, 15, 8000, 0);
		var arc23 = new Arc(2, 3, 3000, 20,    0, 0);
		
		var minTree = new Arc[]{null, arc01, arc02, arc23};
		var maxTree = new Arc[]{null, arc01, arc02, arc13};
		
		Bush bush = new Bush(1, 0);
		
		assertEquals(0, a.findFlowDelta(minTree, maxTree, bush, 3, 0));
	}
}

File: cpp/test/algorithm/FindFlowDeltaTest.cpp
#include "../include.h"

TEST_CASE("findFlowDelta", "[]") {
	Network n(std::vector<std::vector<Network::Arc>>(0), 0);
	Matrix m(0);
	Algorithm a(n, m, 0, 0);

	SECTION("Case from Wikipedia - diamond") {
		Network::Arc arc01(0, 1, 	1,  0, 8000, 0);
		Network::Arc arc02(0, 2, 	1,  0,    0, 0);
		Network::Arc arc13(1, 3, 1000, 15, 8000, 0);
		Network::Arc arc23(2, 3, 3000, 20,    0, 0);

		std::vector<Network::Arc *> minTree{nullptr, &arc01, &arc02, &arc23};
		std::vector<Network::Arc *> maxTree{nullptr, &arc01, &arc02, &arc13};

		Bush bush(1);
		bush.addFlow(0, 10'000);

		REQUIRE(5847 == round(a.findFlowDelta(minTree, maxTree, bush, 3, 0)));
	}

	SECTION("Case from Wikipedia - diamond with tail") {
		Network::Arc arc01(0, 1, 	1,  0,    0, 0);
		Network::Arc arc12(1, 2, 	1,  0, 8000, 0);
		Network::Arc arc13(1, 3, 	1,  0,    0, 0);
		Network::Arc arc24(2, 4, 1000, 15, 8000, 0);
		Network::Arc arc34(3, 4, 3000, 20,    0, 0);

		std::vector<Network::Arc *> minTree{nullptr, &arc01, &arc12, &arc13, &arc34};
		std::vector<Network::Arc *> maxTree{nullptr, &arc01, &arc12, &arc13, &arc24};

		Bush bush(1);
		bush.addFlow(0, 10000);

		REQUIRE(5847 == round(a.findFlowDelta(minTree, maxTree, bush, 4, 1)));
	}

	SECTION("Test maxDeltaX <= μ") {
		Network::Arc arc01(0, 1, 	1,  0, 8000, 0);
		Network::Arc arc02(0, 2, 	1,  0,    0, 0);
		Network::Arc arc13(1, 3, 1000, 15, 8000, 0);
		Network::Arc arc23(2, 3, 3000, 20,    0, 0);

		std::vector<Network::Arc *> minTree{nullptr, &arc01, &arc02, &arc23};
		std::vector<Network::Arc *> maxTree{nullptr, &arc01, &arc02, &arc13};

		Bush bush(1);
		bush.addFlow(0, 1000);

		REQUIRE(1000 == a.findFlowDelta(minTree, maxTree, bush, 3, 0));
	}

	SECTION("Nonconverging case") {
		Network::Arc arc01(0, 1,  17782.7941, 2, 1900, 0);
		Network::Arc arc12(1, 2,  4908.82673, 6,  700, 0);
		Network::Arc arc23(2, 3, 	   10000, 5, 20783.42548485832, 0);
		Network::Arc arc04(0, 4, 	   10000, 5, 2300, 0);
		Network::Arc arc43(4, 3, 13915.78842, 3, 1900, 0);

		std::vector<Network::Arc *> minTree{nullptr, &arc01, &arc12, &arc23, &arc04};
		std::vector<Network::Arc *> maxTree{nullptr, &arc01, &arc12, &arc43, &arc04};

		Bush bush(1);
		bush.addFlow(0, 700);

		REQUIRE(700 == a.findFlowDelta(minTree, maxTree, bush, 3, 0));
	}

	SECTION("maxDeltaX == 0") {
		Network::Arc arc01(0, 1, 	1,  0, 8000, 0);
		Network::Arc arc02(0, 2, 	1,  0,    0, 0);
		Network::Arc arc13(1, 3, 1000, 15, 8000, 0);
		Network::Arc arc23(2, 3, 3000, 20,    0, 0);

		std::vector<Network::Arc *> minTree{nullptr, &arc01, &arc02, &arc23};
		std::vector<Network::Arc *> maxTree{nullptr, &arc01, &arc02, &arc13};

		Bush bush(1);

		REQUIRE(0 == a.findFlowDelta(minTree, maxTree, bush, 3, 0));
	}
}